Stats Revamp: Views and Visitors showing data for the previously selected period or site

staskus opened the pull request:
Fixes #19424

## Description

Views and Visitors showing data for the previously selected period or site

## Root cause

`StatsPeriodStore.getSummary` returns a cached version of summary for a previously selected period

An ideal fix would explore ways to change API of `StatsPeriodStore`, allowing to get a cached summary of a specific period, rather than having one `state.summary` which is changed when the period is changed. The scope of such a change is hard to determine so I decided against it.

## Solution

`SiteStatsInsightsViewModel` always expects to show the chart of `mostRecentData`. Don't set the value of `mostRecentChartData` if it doesn't belong to the most recent period.

## Testing instructions

1. Open Stats
3. See Views & Visitors
4. Open Week >
5. Change period (come back in weeks)
6. Go back to Stats
7. Go back to main menu
8. Select Stats
9. Views & Visitors card will show date for changed period and not for "This week".

## Regression Notes

1. Potential unintended areas of impact

Views & Visitors chart

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Manual testing

3. What automated tests I added (or what prevented me from doing so)

I tried adding tests for `SiteStatsInsightsViewModel` but it would require refactoring to make it testable. Although I would recommend it since it would make making changes much easier and safer.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

staskus committed:
Update mostRecentChardData only if the data belong to the most recent period

Passed PeriodSummary could belong to an older period, however we only want to show the most recent period in the SiteStatsInsightsViewModel. Only setting mostRecentCharData if the passed PeriodSummary belongs to the newest period.

staskus committed:
Remove unnecessary mostRecentChartData updating code

staskus committed:
Merge branch 'trunk' into fix/19424-stats-showing-data-for-previously-selected-period-or-site

guarani commented:
I'm going to review this now. If we can merge this, I think there are no other open Stats Revamp PRs, so things should be in a good state to resume in the future.

guarani reviewed (commented):
Let me see if I understand the issue and how this PR addresses it.

From a user's perspective, the app sometimes shows the wrong data on the Views & Visitors card. This happens after the user has viewed the previous week's data. The wrong data here is actually the data from that previous week. Also, this issue doesn't affect any other card. Do you agree with this?

From a technical perspective, the app looks to be caching data and not checking to ensure the data is from the current period before showing it.

This PR addresses this issue by only using cached data (`periodSummary`) if it's from the same period or a newer period than what the user is looking at. Why don't we only use cached data if it's from the same period (and not from future periods)?

---

I'm not sure if it's related to this, but the "Week" chevron button doesn't respond to taps, so I can't test this PR by looking at previous weeks. I double-checked that it does work with the build from https://github.com/wordpress-mobile/WordPress-iOS/pull/19439. Does it work for you? I wanted to debug this but I'm updating Xcode and it's taking a long time ⌛ 

| <img src="https://user-images.githubusercontent.com/1898325/195631303-a0efc180-4ecc-4c82-9fd6-2d9f795154a6.jpeg" alt="" width="350"> |
| - |

---

I don't follow why `updateMostRecentChartData()` was removed from the code that's for Total Likes and Total Comments. Could you please expand on this?

staskus commented:
Thanks for review and questions, @guarani 👍

Yes, you're understanding the issue correctly.

> This PR addresses this issue by only using cached data (periodSummary) if it's from the same period or a newer period than what the user is looking at. Why don't we only use cached data if it's from the same period (and not from future periods)?

I included a future ending of the period as well, in case we expand the period ending to support "This Week" as it was done in https://github.com/wordpress-mobile/WordPress-iOS/pull/19393. So the inclusion of a future period is not necessary to make the code work in the current situation but I felt that it was more future-proof.

> I'm not sure if it's related to this, but the "Week" chevron button doesn't respond to taps, so I can't test this PR by looking at previous weeks. 

It's something that worked for me during development. 

I installed the app through App Center, enabled the flags, and then "Week >" button didn't respond until I killed the app and went to stats again. 🤔 Maybe it's somehow related to the fact that flags were enabled from the "Debug" menu and some background processes or Stats fetching didn't happen properly.

staskus committed:
Use computed instead of stored variables when FeatureFlag is used

When FeatureFlag is changed after the app is launched, variable values that are computed using FeatureFlags should change

staskus commented:
> I'm not sure if it's related to this, but the "Week" chevron button doesn't respond to taps, so I can't test this PR by looking at previous weeks.

I fixed the issue! The problem was that `FeatureFlag` was used in static stored variables and when `FeatureFlag` changed the value didn't change. Using computed variables instead. I went through all the feature flags and I didn't find any such usage of feature flags outside Stats feature.

guarani commented:
Turned this comment into a review below https://github.com/wordpress-mobile/WordPress-iOS/pull/19428#pullrequestreview-1142505186

guarani reviewed (commented):
> I included a future ending of the period as well, in case we expand the period ending to support "This Week" as it was done in https://github.com/wordpress-mobile/WordPress-iOS/pull/19393. So the inclusion of a future period is not necessary to make the code work in the current situation but I felt that it was more future-proof.

Thanks for sharing that. I didn't realize that the "insights" Views & Visitors card was showing the last 7 days while the "week" Views & Visitors card was showing data for this week.

Nice work catching the feature flag issue. Does this mean the "Weeks >" button wasn't working whenever we visited the old Stats first and then turned on the feature flag?

> I don't follow why updateMostRecentChartData() was removed from the code that's for Total Likes and Total Comments. Could you please expand on this?

I'm still not sure about this part, could you please explain this change?

staskus commented:
> Nice work catching the feature flag issue. Does this mean the "Weeks >" button wasn't working whenever we visited the old Stats first and then turned on the feature flag?

Yes. Also "Latest Post Summary" cell would never turn into a new appearance since its cell type was set once.

> I don't follow why updateMostRecentChartData() was removed from the code that's for Total Likes and Total Comments. Could you please expand on this?

That code there doesn't serve any purpose. Neither `periodSummary` nor `mostRecentChartData` is used when creating totalLikes and totalComments row. 

Previously, mostRecentChartData **was** used, in order to pass `sparklineData` into `StatsTotalInsightsData` initialization method.

<img width="956" alt="image" src="https://user-images.githubusercontent.com/4062343/195870872-88e570c5-fa0f-4e1f-9190-ba80947bf20d.png">

guarani reviewed (approved):
> Yes. Also "Latest Post Summary" cell would never turn into a new appearance since its cell type was set once.

So we've been seeing the old "Latest Post Summary" even when the feature flags were on? I don't recall seeing the old style.

| Old | New |
| - | - |
| <img src="https://user-images.githubusercontent.com/1898325/195944840-cb6e29de-2135-4bef-9fe1-6645f6825f01.jpeg" alt="" width="350"> | <img src="https://user-images.githubusercontent.com/1898325/195944833-4418e374-8fa0-41ac-8b24-4dca2c21bbd0.jpeg" alt="" width="350"> |


> That code there doesn't serve any purpose. Neither `periodSummary` nor `mostRecentChartData` is used when creating totalLikes and totalComments row.
> 
> Previously, mostRecentChartData **was** used, in order to pass `sparklineData` into `StatsTotalInsightsData` initialization method.

Gotcha, thanks for explaining.

I tested the latest changes here in https://github.com/wordpress-mobile/WordPress-iOS/pull/19428#pullrequestreview-1142505186 and there have been no changes since, so I'm happy to approve. If this was going to production, I might consider merging this since the release is being cut Monday. Since it's still behind a feature flag, I'll stick to the repo norms and leave you to merge this Monday.

mokagio commented:
> If this was going to production, I might consider merging this since the release is being cut Monday. Since it's still behind a feature flag, I'll stick to the repo norms and leave you to merge this Monday.

@guarani made a good observation here. Still, I'll take over merging from @staskus because it's Monday (for me, at least) and I'm about to run the code freeze for 21.0.

Even if the feature itself is not enabled, I think it's still useful to get the code itself in the release branch, to make sure it works well, even if just at the integration point.

staskus commented:
> So we've been seeing the old "Latest Post Summary" even when the feature flags were on? I don't recall seeing the old style.

Yes, until we would kill and relaunch the app **or** if we would enable the feature flag from code and not from settings.
